modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/WeightedProspect.scala

```diff
@@ -0,0 +1,51 @@
+package org.tessellation.sdk.infrastructure.snapshot
+
+import cats.MonadThrow
+import cats.data.Chain
+import cats.effect.std.Random
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.trust.TrustValueRefined
+import org.tessellation.sdk.infrastructure.snapshot.WeightedSampling.{
+  ProbabilityRefinement,
+  Weight,
+  WeightRefinement,
+  sample => weightedSample
+}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.refineV
+import eu.timepit.refined.types.numeric.PosInt
+
+object WeightedProspect {
+
+  def sample[F[_]: MonadThrow: Random, A](candidates: Map[A, TrustValueRefined], sampleSize: PosInt): F[List[A]] = {
+    val filtered = candidates.view
+      .mapValues(score => refineV[WeightRefinement](score.value))
+      .collect {
+        case (key, Right(refinedScore)) =>
+          key -> refinedScore
+      }
+      .toMap
```

Comment:
If you would store scores as `Weight` type in the `TrustValueRefined`, you won't need this multiple remaps and collects

Reply:
Trust scores can be from `-1` to `1`, inclusive on both ends. We want to store them as such. That's `TrustValueRefined`. However, when selecting a peer based on trust values, we only want to select peers with non-negative trust scores. So, `Weight` is `0` exclusive to `1` inclusive.